components/script/dom/node.rs

```diff
@@ -876,12 +946,29 @@ impl Node {
         self.owner_doc.set(Some(document));
     }
 
+    pub fn containing_shadow_root(&self) -> Option<DomRoot<ShadowRoot>> {
+        // NodeRareData contains the shadow root the node belongs to,
```

Comment:
Mind explaining how can a shadow root belong to another shadow root? Can we instead make a `shadow_root.set_containing_shadow_root(shadow_root)` in `AttachShadow`, and then remove this and the similar special-cases? I think that'd be nicer.